hypertrace-core-graphql-span-schema/src/main/java/org/hypertrace/core/graphql/span/joiner/DefaultSpanJoinerBuilder.java

```diff
@@ -91,34 +91,47 @@ private class DefaultSpanJoiner implements SpanJoiner {
     private final List<SelectedField> selectedFields;
 
     @Override
-    public <T> Single<Map<T, Span>> joinSpans(
+    public <T> Single<Map<T, Collection<Span>>> joinSpans(
         Collection<T> joinSources, SpanIdGetter<T> spanIdGetter) {
       return this.buildSourceToIdMap(joinSources, spanIdGetter).flatMap(this::joinSpans);
     }
 
-    private <T> Single<Map<T, Span>> joinSpans(Map<T, String> sourceToSpanIdMap) {
-      return this.buildSpanRequest(sourceToSpanIdMap)
+    private <T> Single<Map<T, Collection<Span>>> joinSpans(
+        Map<T, Collection<String>> sourceToSpanIdsMap) {
+      return this.buildSpanRequest(sourceToSpanIdsMap)
           .flatMap(spanDao::getSpans)
           .map(this::buildSpanIdToSpanMap)
-          .map(spanIdToSpanMap -> buildSourceToSpanMap(sourceToSpanIdMap, spanIdToSpanMap));
+          .map(spanIdToSpanMap -> buildSourceToSpansMap(sourceToSpanIdsMap, spanIdToSpanMap));
     }
 
-    private <T> Map<T, Span> buildSourceToSpanMap(
-        Map<T, String> sourceToSpanIdMap, Map<String, Span> spanIdToSpanMap) {
-      return sourceToSpanIdMap.entrySet().stream()
-          .filter(entry -> spanIdToSpanMap.containsKey(entry.getValue()))
+    private <T> Map<T, Collection<Span>> buildSourceToSpansMap(
+        Map<T, Collection<String>> sourceToSpanIdsMap, Map<String, Span> spanIdToSpanMap) {
+      return sourceToSpanIdsMap.entrySet().stream()
           .collect(
               Collectors.toUnmodifiableMap(
-                  Entry::getKey, entry -> spanIdToSpanMap.get(entry.getValue())));
+                  Entry::getKey,
+                  entry -> buildSpansCollectionForEachSource(entry.getValue(), spanIdToSpanMap)));
+    }
+
+    private Collection<Span> buildSpansCollectionForEachSource(
+        Collection<String> spanIds, Map<String, Span> spanIdToSpanMap) {
+      return spanIds.stream()
+          .filter(spanIdToSpanMap::containsKey)
+          .distinct()
```

Comment:
If the caller hands you the same ID twice, why would we only return it to them once? The result list should match the input list - it's on the caller to dedupe, if desired. In the internal code like building the filter for the span query - there it makes sense to dedupe, but we want the response to the caller to reflect the request.